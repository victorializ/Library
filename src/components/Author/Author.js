import React from 'react';

import { List } from 'semantic-ui-react';

const Author = (
  {
    authorId, 
    firstName, 
    lastName,
    onAuthorSelect
  }
) => {
  return (
    <List.Item 
      key={authorId} 
    >
      <List.Content 
        className='authors__list-item'
        onClick={() => onAuthorSelect(authorId)}
      >
        <List.Header>
          {firstName} {lastName}
        </List.Header>
      </List.Content>
    </List.Item>
  );
};

export { Author };