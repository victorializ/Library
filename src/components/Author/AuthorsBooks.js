import React from 'react';

import { BooksList, ErrorMessage } from '../../components';
import { constants } from '../../i18n';

const AuthorsBooks = ({
    data: {
        bookAuthors: books
    }
}) =>
  books.length ? 
    <BooksList data={books.map(({book}) => book)} />
    : 
    <ErrorMessage text={constants.nobooks} />

export { AuthorsBooks };