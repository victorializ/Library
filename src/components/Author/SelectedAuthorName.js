import React from 'react';

import { Segment } from 'semantic-ui-react';

const SelectedAuthorName = (
    {
        data: {
            firstName = '', 
            lastName = ''
        }
    }
) =>
  (firstName && lastName) && 
    <Segment className='authors__name'>{`${firstName} ${lastName}`}</Segment>

export { SelectedAuthorName };