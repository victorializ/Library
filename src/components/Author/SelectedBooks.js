import React from 'react';

import { BooksList, ErrorMessage } from '../../components';
import { constants } from '../../i18n';

const SelectedBooks = ({
    data: {
        bookAuthors: books
    }
}) =>
  books.length ? 
    <BooksList 
        data={books.map(({book}) => book)} 
    />
    : 
    <ErrorMessage text={constants.nobooks} />

export { SelectedBooks as SelectedAuthorsBooks };