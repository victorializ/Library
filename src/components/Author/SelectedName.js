import React from 'react';

import { Segment } from 'semantic-ui-react';

const SelectedName = (
    {
        data: {
            firstName = '', 
            lastName = ''
        }
    }
) =>
  (firstName && lastName) && 
    <Segment className='authors__name'>{`${firstName} ${lastName}`}</Segment>

export { SelectedName as SelectedAuthorName };