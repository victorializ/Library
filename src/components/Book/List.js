import React  from 'react';

import { Book } from './Book';
import { ItemsList } from '../ItemsList';

const List = ({data}) => 
    <ItemsList 
        data={data} 
        Component={Book}
    />

export { List as BooksList };    
