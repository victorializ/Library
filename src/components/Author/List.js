import React from 'react';

import { List } from 'semantic-ui-react';
import { Author } from './Author';

const AuthorsList = (
  {
    data,
    onAuthorSelect
  }) => 
  <List inverted divided>
    { data.map(
      ({
        authorId, 
        firstName, 
        lastName
      }) => 
      <Author 
          key={authorId}
          authorId={authorId}
          firstName={firstName}
          lastName={lastName}
          onAuthorSelect={onAuthorSelect}
          className='authors__list-item'
      />
    )}
  </List>

export { AuthorsList };