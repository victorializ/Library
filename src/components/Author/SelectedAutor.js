import React from 'react';

import { getAuthor } from '../../services/http-client';
import { useRequest } from '../../hooks';
import { LoadingComponent } from '../../components';
import { SelectedAuthorName } from './SelectedName';
import { SelectedAuthorsBooks } from './SelectedBooks';

const SelectedAuthor = ({authorId}) => {
  const response = useRequest(getAuthor, [authorId]);
  return (
    <div className='authors--selected'>
      <div className='authors--selected__name'>
        <LoadingComponent {...response}>
          <SelectedAuthorName data={response.data} /> 
        </LoadingComponent>
      </div>
      <div className='authors--selected__books'>
        <LoadingComponent {...response}>
          <SelectedAuthorsBooks data={response.data} />
        </LoadingComponent>
      </div>
    </div>
  ); 
};

export { SelectedAuthor };