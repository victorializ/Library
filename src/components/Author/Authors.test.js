import React from 'react';
import { shallow } from 'enzyme';

import { Author } from './Author';
import { AuthorsList } from './List';
import { SelectedAuthor } from './SelectedAutor';
import { SelectedAuthorsBooks } from './SelectedBooks';
import { SelectedAuthorName } from './SelectedName'; 

describe('Authors', () => {
    it('renders author component correctly', () => {
        expect(shallow(<Author />)).toMatchSnapshot();
    }); 
    it('triggers function on author click', () => {
        const id = 1;
        const handleClick = jest.fn();
        const wrapper = shallow(<Author authorId={id} onAuthorSelect={handleClick} />);
        wrapper.find('.authors__list-item').simulate('click');
        expect(handleClick.mock.calls[0][0]).toBe(id);
    });  
    it('renders authors list component correctly', () => {
        expect(shallow(<AuthorsList data={[]}/>)).toMatchSnapshot();
    });
    it('renders list of two authors', () => {
        const data = [{}, {}];
        const wrapper = shallow(<AuthorsList data={data}/>);
        expect(wrapper.find('.authors__list-item').length).toBe(data.length);
    });
    it('renders selected author component correctly', () => {
        expect(shallow(<SelectedAuthor />)).toMatchSnapshot();
    });
    it('renders selected books component correctly', () => {
        expect(shallow(<SelectedAuthorsBooks data={{bookAuthors: []}} />)).toMatchSnapshot();
    });
    it('renders list of two books', () => {
        const data = {bookAuthors: [{}, {}]}
        const wrapper = shallow(<SelectedAuthorsBooks data={data}/>);
        expect(wrapper.find('.books__list').exists()).toBeTruthy();
    });
    it('renders selected name component correctly', () => {
        expect(shallow(<SelectedAuthorName data={{}} />)).toMatchSnapshot();
    });
    it('renders selected authors name when available', () => {
        const data = {firstName: 'FN', lastName: 'LN'};
        const wrapper = shallow(<SelectedAuthorName data={data} />);
        expect(wrapper.find('.authors__name').exists()).toBeTruthy();
    });
    it('not renders selected authors name when not available', () => {
        const data = {};
        const wrapper = shallow(<SelectedAuthorName data={data} />);
        expect(wrapper.find('.authors__name').exists()).toBeFalsy();
    });
});