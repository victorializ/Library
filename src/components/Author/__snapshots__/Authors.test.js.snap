// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authors renders author component correctly 1`] = `
<ListItem>
  <ListContent
    className="authors__list-item"
    onClick={[Function]}
  >
    <ListHeader>
       
    </ListHeader>
  </ListContent>
</ListItem>
`;

exports[`Authors renders authors list component correctly 1`] = `
<List
  divided={true}
  inverted={true}
/>
`;

exports[`Authors renders selected author component correctly 1`] = `
<div
  className="authors--selected"
>
  <div
    className="authors--selected__name"
  >
    <LoadingComponent
      data={null}
      error={
        Object {
          "message": null,
        }
      }
      loading={true}
    >
      <SelectedName
        data={null}
      />
    </LoadingComponent>
  </div>
  <div
    className="authors--selected__books"
  >
    <LoadingComponent
      data={null}
      error={
        Object {
          "message": null,
        }
      }
      loading={true}
    >
      <SelectedBooks
        data={null}
      />
    </LoadingComponent>
  </div>
</div>
`;

exports[`Authors renders selected books component correctly 1`] = `
<ErrorMessage
  text="No books available!"
/>
`;

exports[`Authors renders selected name component correctly 1`] = `""`;
